internal/api/impl/v1/dashboard/schemas/validator.go

```diff
@@ -127,70 +117,121 @@ func (v *validator) Validate(panels map[string]json.RawMessage) error {
+	chartsPath := filepath.Join(v.schemasConf.Path, v.schemasConf.ChartsFolder)
 
-	files, err := os.ReadDir(pluginsPath)
+	charts, err := os.ReadDir(chartsPath)
 	if err != nil {
-		logrus.WithError(err).Errorf("Not able to read from schemas dir %s", pluginsPath)
+		logrus.WithError(err).Errorf("Not able to read from charts dir %s", chartsPath)
 		return
 	}
 
-	// remove any previous item before proceeding
+	// reset the validator's schemas list before proceeding
 	v.schemas.Range(func(key interface{}, value interface{}) bool {
 		v.schemas.Delete(key)
 		return true
 	})
 
-	// process each schema plugin to convert it into a CUE Value
-	// for each schema we check that it meets the default specs we expect for any panel, otherwise we dont include it
-	for _, file := range files {
-		if !file.IsDir() {
-			logrus.Warningf("Plugin %s is not a folder", file.Name())
+	// process each chart plugin to convert it into a CUE Value
+	// for each chart we check that its schema meets the default specs we expect for any chart, otherwise we dont include it
+	for _, chart := range charts {
+		if !chart.IsDir() {
+			logrus.Warningf("Chart plugin %s is not a folder", chart.Name())
 			continue
 		}
 
-		schemaPath := filepath.Join(pluginsPath, file.Name())
-
-		// load the cue files into build.Instances slice
-		buildInstances := load.Instances([]string{}, &load.Config{Dir: schemaPath})
-		// we strongly assume that only 1 buildInstance should be returned (corresponding to the main definition like #panel), otherwise we skip it
-		// TODO can probably be improved
-		if len(buildInstances) != 1 {
-			logrus.Errorf("The number of build instances for %s is != 1, skipping this schema", schemaPath)
+		// schemaFiles will be the list of all CUE files to consider to build the final chart schema
+		var schemaFiles []string
+		schemaPath := filepath.Join(chartsPath, chart.Name())
+		// - first, put the base chart def
+		schemaFiles = append(schemaFiles, filepath.Join(v.schemasConf.Path, baseDefFile))
+		// - then, retrieve & add all the CUE files from this chart
+		chartFiles, err := listCueFiles(schemaPath)
+		if err != nil {
+			logrus.WithError(err).Errorf("Not able to retrieve the chart's CUE files from dir %s", schemaPath)
 			continue
 		}
-		buildInstance := buildInstances[0]
-
-		// check for errors on the instances (these are typically parsing errors)
-		if buildInstance.Err != nil {
-			logrus.WithError(buildInstance.Err).Errorf("Error retrieving schema for %s, skipping this schema", schemaPath)
+		schemaFiles = append(schemaFiles, chartFiles...)
+		// - then, add all the query types available
+		queriesPath := filepath.Join(v.schemasConf.Path, v.schemasConf.QueriesFolder)
+		queriesFiles, err := listCueFiles(queriesPath)
+		if err != nil {
+			logrus.WithError(err).Errorf("Not able to retrieve the queries CUE files from dir %s", queriesPath)
 			continue
 		}
+		schemaFiles = append(schemaFiles, queriesFiles...)
```

Comment:
L148-L153 and L155 - L161 are the same. Times to create a nice function to call it twice to load the charts and the queries :)